core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,348 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.base.VerifyException;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.DataSize;
+import io.trino.execution.TaskId;
+import io.trino.execution.buffer.SerializedPage;
+import io.trino.spi.TrinoException;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.util.concurrent.Futures.nonCancellationPropagating;
+import static io.trino.operator.RetryPolicy.QUERY;
+import static io.trino.operator.RetryPolicy.TASK;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class DeduplicationExchangeClientBuffer
+        implements ExchangeClientBuffer
+{
+    private final Executor executor;
+    private final long bufferCapacityInBytes;
+    private final RetryPolicy retryPolicy;
+
+    private final SettableFuture<Void> blocked = SettableFuture.create();
+    @GuardedBy("this")
+    private final Set<TaskId> allTasks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreTasks;
+    @GuardedBy("this")
+    private final Set<TaskId> finishedTasks = new HashSet<>();
+    @GuardedBy("this")
+    private final Map<TaskId, Throwable> failedTasks = new HashMap<>();
+    @GuardedBy("this")
+    private boolean inputFinished;
+    @GuardedBy("this")
+    private Throwable failure;
+
+    @GuardedBy("this")
+    private final ListMultimap<TaskId, SerializedPage> pageBuffer = ArrayListMultimap.create();
+    @GuardedBy("this")
+    private Iterator<SerializedPage> pagesIterator;
+    @GuardedBy("this")
+    private volatile long bufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private volatile long maxBufferRetainedSizeInBytes;
+
+    @GuardedBy("this")
+    private boolean closed;
+
+    public DeduplicationExchangeClientBuffer(Executor executor, DataSize bufferCapacity, RetryPolicy retryPolicy)
+    {
+        this.executor = requireNonNull(executor, "executor is null");
+        this.bufferCapacityInBytes = requireNonNull(bufferCapacity, "bufferCapacity is null").toBytes();
+        this.retryPolicy = requireNonNull(retryPolicy, "retryPolicy is null");
```

Comment:
check `retryPolicy` is one of allowed values. I would reject anything but `QUERY` at this stage already in the constructor; but I guess `//TODO+throw` in the code below are also fine. Still I think it does not make sense to have `NONE` here.

Reply:
Added a check for NONE